starfish/util/synthesize.py

```diff
@@ -0,0 +1,202 @@
+
+    image_with_background = image + poisson(p['N_photon_background'], size=image.shape)
+    filtered = array([gaussian(im, p['psf']) for im in image_with_background])
+    filtered = filtered * p['detection_efficiency'] + normal(scale=p['N_background_electrons'], size=filtered.shape)
+    signal = np.array([(x / p['graylevel']).astype(int).clip(0, 2 ** p['bits']) for x in filtered])
+
+    def select_uint_dtype(array):
+        """choose appropriate dtype based on values of an array"""
+        max_val = np.max(array)
+        for dtype in [np.uint8, np.uint16, np.uint32, np.uint64]:
+            if max_val <= dtype(-1):
+                return array.astype(dtype)
+        raise ValueError('value exceeds dynamic range of largest numpy type')
+
+    corrected_signal = select_uint_dtype(signal)
+    rescaled_signal: np.ndarray = rescale_intensity(corrected_signal)
+
+    # set up the tile set
+    image_data = TileSet(
+        {Coordinates.X, Coordinates.Y, Indices.HYB, Indices.CH, Indices.Z},
+        {
+            Indices.HYB: NUM_HYB,
+            Indices.CH: NUM_CH,
+            Indices.Z: NUM_Z,
+        },
+        default_tile_shape=(Y, X),
+    )
+
+    # fill the TileSet
+    experiment_indices = list(product(range(NUM_HYB), range(NUM_CH), range(NUM_Z)))
+    for i, (hyb, ch, z) in enumerate(experiment_indices):
+
+        tile = Tile(
+            {
+                Coordinates.X: (0.0, 0.001),
+                Coordinates.Y: (0.0, 0.001),
+                Coordinates.Z: (0.0, 0.001),
+            },
+            {
+                Indices.HYB: hyb,
+                Indices.CH: ch,
+                Indices.Z: z,
+            }
+        )
+        tile.numpy_array = rescaled_signal[i]
+
+        image_data.add_tile(tile)
+
+    data_stack = ImageStack(image_data)
+
+    # make a max projection and pretend that's the dots image, which we'll create another ImageStack for this
+    dots_data = TileSet(
+        {Coordinates.X, Coordinates.Y, Indices.HYB, Indices.CH, Indices.Z},
+        {
+            Indices.HYB: 1,
+            Indices.CH: 1,
+            Indices.Z: 1,
+        },
+        default_tile_shape=(Y, X),
+    )
+    tile = Tile(
+        {
+            Coordinates.X: (0.0, 0.001),
+            Coordinates.Y: (0.0, 0.001),
+            Coordinates.Z: (0.0, 0.001),
+        },
+        {
+            Indices.HYB: 0,
+            Indices.CH: 0,
+            Indices.Z: 0,
+        }
+    )
+
+    tile.numpy_array = np.max(rescaled_signal, 0)
+
+    dots_data.add_tile(tile)
+    dots_stack = ImageStack(dots_data)
+
+    # TODO can we mock up a nuclei image somehow?
+
+    # put the data together into a top-level Stack
+    results = Stack.from_data(data_stack, aux_dict={'dots': dots_stack})
+
+    # make the codebook(s)
+    codebook = []
+    for _, code_record in spots.iterrows():
+        codeword = []
+        for code_value, (hyb, ch, z) in zip(code_record['barcode'], experiment_indices):
+            if code_value == 1:
```

Comment:
if code_value != 0:

Reply:
Ah, fair point. I still have no idea where that's going to come up. :-)

Reply:
Just being prepared. :)